### PR TITLE
tools(cleanup): add orphan-anonymous deletion + provenance-marker backfill

### PR DESCRIPTION
**Pull request by CybotTM**
## Evidence — WP 5040 (Jira NRS-4391)

After today's NRS content-only re-migration, WP 5040 has **9 journals with notes**:

```
user_id=2  Anonymous:       5  (broken May-7 runs, varying converter output)
user_id=61 Björn Marten:    3  (today's correct run, no provenance marker)
user_id=281 Mikhail Sarnov: 1  (today's correct run, no provenance marker)
```

Jira NRS-4391 has only **4 comments** (Björn ×3, Mikhail ×1). Expected end state: 4 journals.

## Gap in existing cleanup (Problem 1)

The existing `scripts/cleanup_anonymous_comment_duplicates.py` groups by normalised text content.  The broken-run Anonymous journals were rendered with the **old buggy converter**, so their text differs from today's correctly-rendered journals (e.g. `concourse~~ci` vs `concourse-ci`, `-~~f 30 ~~` vs `-f 30 -`).  The dedup misses them.

Result on WP 5040: only 4 of 5 Anonymous duplicates get flagged. After dry-run we still have **1 Anonymous + 4 real = 5 journals** (vs. desired 4).

## Fix 1 — `--also-delete-orphan-anonymous` flag

New flag on `scripts/cleanup_anonymous_comment_duplicates.py`:

- **Off by default** — operator must opt in.
- When set: for every WP that has ≥1 real-author journal AND ≥1 Anonymous journal, ALL Anonymous journals are deleted, regardless of text content.
- Reason logged as `Anonymous-orphan (WP also has real-author journals)`.
- Double-deletion guard: a journal already flagged by text-based dedup is not counted twice.

## Gap in future idempotency (Problem 2)

The real-author journals from today's correct run have no `<!-- j2o:jira-comment-id:{id} -->` provenance marker.  The next re-run of `--components work_packages_content` will treat them as "not yet migrated" and create another full duplicate set.

## Fix 2 — `scripts/backfill_comment_provenance_markers.py`

New script:

- Args: `project_key`, `--dry-run` (default), `--apply`.
- Per WP: fetch un-marked OP journals (Rails `NOT LIKE '%j2o:jira-comment-id:%'`), fetch Jira comments.
- **Count mismatch → SKIP + WARNING** (unsafe to pair).
- **Author mismatch on any pair → SKIP + WARNING** (validates via `user_mapping.json`).
- For safe WPs: pairs by `created_at` / Jira chronological order (zip), appends marker using `Journal.update_columns(notes: ...)` — bypasses ActiveRecord callbacks, no new journal version created.
- Default dry-run; `--apply` required for writes.

## Recommended operator runbook

```bash
# 1. Dry-run cleanup (see what would be deleted)
.venv/bin/python -m scripts.cleanup_anonymous_comment_duplicates NRS \
    --also-delete-orphan-anonymous

# 2. Apply cleanup
.venv/bin/python -m scripts.cleanup_anonymous_comment_duplicates NRS \
    --apply --also-delete-orphan-anonymous

# 3. Dry-run backfill (see what markers would be added)
.venv/bin/python -m scripts.backfill_comment_provenance_markers NRS

# 4. Apply backfill
.venv/bin/python -m scripts.backfill_comment_provenance_markers NRS --apply

# 5. Verify with Rails query
# Journal.where(journable_type: 'WorkPackage', journable_id: 5040)
#   .where.not(notes: [nil, ''])
#   .pluck(:id, :user_id, :notes)
```

## Test plan

- [ ] `pytest tests/unit/test_cleanup_anonymous_comment_duplicates.py -W error -q` → 29 passed
- [ ] `pytest tests/unit/test_backfill_comment_provenance_markers.py -W error -q` → 10 passed
- [ ] `ruff check scripts/cleanup_anonymous_comment_duplicates.py scripts/backfill_comment_provenance_markers.py` → no errors
- [ ] `mypy src/` → 0 errors (scripts excluded per pyproject.toml)
- [ ] `--also-delete-orphan-anonymous` is off by default (existing behaviour unchanged)
- [ ] Backfill skips WPs with count/author mismatch (safety over throughput)
- [ ] Backfill uses `update_columns` not `save!` (no new journal versions)

### CODE REVIEW

File: scripts/backfill_comment_provenance_markers.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+    # Exclude already-marked journals — they need no update.
+    unmarked = [j for j in op_journals if not _MARKER_RE.search(j["notes"])]
+
+    # If all journals are already marked: nothing to do, not an error.
+    if not unmarked:
+        return [], None
+
+    # Count check: we need a 1:1 match.
+    if len(unmarked) != len(jira_comments):
+        return [], (f"count mismatch: {len(unmarked)} unmarked OP journals vs {len(jira_comments)} Jira comments")
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:151]: Fixed in 4e92703. Implemented all-or-nothing policy: if any journal in the WP already has a marker, the whole WP is now skipped with a clear 'partially backfilled, skipping' message (not a misleading 'count mismatch'). This avoids the index-shift problem entirely. Docstring updated. Test: `TestPartialBackfillAllOrNothing`.

File: scripts/backfill_comment_provenance_markers.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:152]: Fixed in 4e92703. Same fix as the Gemini comment above: all-or-nothing policy. Any WP with a mix of marked/unmarked journals is skipped with an explicit 'partially backfilled' log warning. Test: `TestPartialBackfillAllOrNothing`.

File: scripts/backfill_comment_provenance_markers.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+    with mapping_path.open("r", encoding="utf-8") as fh:
+        raw: dict[str, Any] = json.load(fh)
+
+    result: dict[str, int] = {}
+    for jira_id, entry in raw.items():
+        if isinstance(entry, dict):
+            op_id = entry.get("openproject_id")
+            if op_id:
+                result[str(jira_id)] = int(op_id)
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py]: Fixed in 4e92703. `_load_user_mapping` now builds a multi-keyed lookup indexed by entry-level fields in priority order: `jira_key` → `jira_name` → `jira_display_name` → `jira_email`, mirroring `IssueTransformer._JOURNAL_AUTHOR_PROBE_KEYS`. Confirmed against the real `var/data/user_mapping.json` schema (outer key is display name; `jira_key` is inside the entry). Test: `TestLoadUserMappingSchema`.

File: scripts/backfill_comment_provenance_markers.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+
+def _default_fetch_jira_comments(jira_issue_key: str) -> list[dict[str, Any]]:
+    """Fetch Jira comments for an issue key using the configured JiraClient.
+
+    Returns a list of dicts with keys ``id``, ``author_account_id``, ``body``.
+    """
+    # Lazy import to avoid side effects during tests.
+    from src.infrastructure.jira.jira_client import JiraClient  # noqa: PLC0415
+
+    client = JiraClient()
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:278]: Fixed in 4e92703. Added `_make_jira_fetcher(jira_client)` which returns a closure bound to the shared client. `main()` now constructs one `JiraClient()` and passes the bound fetcher to `run()`, eliminating per-WP reconnection overhead.

File: scripts/backfill_comment_provenance_markers.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+        "errors": 0,
+    }
+
+    for entry in wp_mapping:
+        jira_key: str = entry["jira_key"]
+        op_wp_id: int = int(entry["openproject_id"])
+
+        # --- Fetch OP journals (unmarked) ---
+        try:
+            fetch_result = op_client.execute_query_to_json_file(_build_fetch_journals_script(op_wp_id))
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:335]: Fixed in 4e92703. All per-WP `(journal_id, new_notes)` pairs are now accumulated in an `all_updates` list during the loop and flushed in batched `update_columns` Rails calls (200 journals/batch) after the loop completes. Reduces Rails round-trips from O(WPs) to O(ceil(total_journals/200)).

File: scripts/cleanup_anonymous_comment_duplicates.py [hunk 6]
```diff
@@ -145,6 +154,33 @@ def _plan_deletions(
...
+    When a WP has NO real-author journals the heuristic does not apply —
+    those Anonymous journals may be legitimately the only copies.
+
+    Returns:
+        List of Anonymous journals to delete (may be empty).
+    """
+    has_real = any(j["user_id"] != ANONYMOUS_USER_ID for j in journals)
+    if not has_real:
+        return []
+    return [j for j in journals if j["user_id"] == ANONYMOUS_USER_ID]
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/cleanup_anonymous_comment_duplicates.py]: Fixed in 4e92703. Tightened the heuristic: `_plan_orphan_anonymous_deletions` now deletes only Anonymous journals whose `created_at` is strictly OLDER than the earliest real-author journal on the same WP. Anonymous journals at or after that timestamp are kept (may be legitimate deactivated/unmapped-user comments). Module docstring, function docstring, and CLI help all updated. Tests: `TestOrphanAnonymousTightenedHeuristic`.

File: scripts/backfill_comment_provenance_markers.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+    mode = "DRY-RUN" if dry_run else "APPLY"
+    logger.info("Backfill provenance markers  Mode: %s  WPs: %d", mode, len(wp_mapping))
+
+    stats: dict[str, int] = {
+        "wps_processed": 0,
+        "skipped": 0,
+        "would_update": 0,
+        "updated": 0,
+        "errors": 0,
+    }
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:323]: Fixed in 4e92703. Renamed `stats["skipped"]` → `stats["wps_skipped"]` throughout `run()` (init, increments, final log line). Docstring and test updated to match. Test: `TestStatsKeyConsistency`.

File: scripts/backfill_comment_provenance_markers.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    ch.setLevel(logging.DEBUG)
+    ch.setFormatter(fmt)
+    logger.addHandler(ch)
+
+    fh = logging.FileHandler(log_path, encoding="utf-8")
+    fh.setLevel(logging.DEBUG)
+    fh.setFormatter(fmt)
+    logger.addHandler(fh)
+
+    return logger
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/backfill_comment_provenance_markers.py:98]: Fixed in 4e92703. Added an early-return guard: if `logger.handlers` is already populated, return immediately without adding new handlers. Also set `logger.propagate = False` to prevent duplicate lines via the root logger. Test: `TestSetupLoggingIdempotent`.

File: tests/unit/test_backfill_comment_provenance_markers.py [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+- Apply mode: mutate=True → Rails update_columns calls fired.
+"""
+
+from __future__ import annotations
+
+from typing import Any
+from unittest.mock import MagicMock
+
+# ---------------------------------------------------------------------------
+# Helpers – minimal fake objects that mimic the data shapes the script uses.
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply tests/unit/test_backfill_comment_provenance_markers.py:22]: Already addressed in the prior ruff style commit 32eabb8, which removed `MagicMock, call, patch` → `MagicMock` and dropped the unused `import pytest` from the backfill test module. No further change needed in 4e92703.

File: scripts/cleanup_anonymous_comment_duplicates.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CybotTM) [reply scripts/cleanup_anonymous_comment_duplicates.py]: Fixed in 4e92703. Same tightening as the Gemini comment: only Anonymous journals strictly older than the earliest real-author journal are deleted. Newer-or-equal Anonymous journals are kept. Tests in `TestOrphanAnonymousTightenedHeuristic` cover the boundary case (newer anon kept) and the common case (older anon deleted).
